src/dotty/tools/dotc/Run.scala

```diff
@@ -56,26 +64,103 @@ class Run(comp: Compiler)(implicit ctx: Context) {
     val phases = ctx.squashPhases(ctx.phasePlan,
       ctx.settings.Yskip.value, ctx.settings.YstopBefore.value, ctx.settings.YstopAfter.value, ctx.settings.Ycheck.value)
     ctx.usePhases(phases)
+    var lastPrintedTree: PrintedTree = NoPrintedTree
     for (phase <- ctx.allPhases)
       if (!ctx.reporter.hasErrors) {
         val start = System.currentTimeMillis
         units = phase.runOn(units)
-        def foreachUnit(op: Context => Unit)(implicit ctx: Context): Unit =
-          for (unit <- units) op(ctx.fresh.setPhase(phase.next).setCompilationUnit(unit))
-        if (ctx.settings.Xprint.value.containsPhase(phase))
-          foreachUnit(printTree)
+        if (ctx.settings.Xprint.value.containsPhase(phase)) {
+          for (unit <- units) {
+            lastPrintedTree =
+              printTree(ctx.fresh.setPhase(phase.next).setCompilationUnit(unit), lastPrintedTree)
+          }
+        }
         ctx.informTime(s"$phase ", start)
       }
     if (!ctx.reporter.hasErrors) Rewrites.writeBack()
   }
 
-  private def printTree(ctx: Context) = {
+  private sealed trait PrintedTree
+  private final case class SomePrintedTree(phase: String, tree: String) extends PrintedTree
+  private object NoPrintedTree extends PrintedTree
+
+  private def printTree(ctx: Context, last: PrintedTree): PrintedTree = {
     val unit = ctx.compilationUnit
     val prevPhase = ctx.phase.prev // can be a mini-phase
     val squashedPhase = ctx.squashed(prevPhase)
+    val treeString = unit.tpdTree.show(ctx)
+
+    ctx.echo(s"result of $unit after $squashedPhase:")
+
+    last match {
+      case SomePrintedTree(phase, lastTreeSting) if lastTreeSting != treeString =>
+        ctx.echo(mkColoredDiffTree(treeString, lastTreeSting))
+        SomePrintedTree(squashedPhase.toString, treeString)
+
+      case SomePrintedTree(phase, lastTreeSting) =>
+        ctx.echo("  Unchanged since " + phase)
+        last
+
+      case NoPrintedTree =>
+        ctx.echo(treeString)
+        SomePrintedTree(squashedPhase.toString, treeString)
+    }
+  }
+
+  private def mkColoredDiffTree(treeString: String, lastTreeSting: String): String = {
```

Comment:
it would be nice to factor this out to a module reusable from other parts of compiler. 
This logic doesn't seem specific to `Run`.